amr-wind/physics/ForestDrag.cpp

```diff
@@ -0,0 +1,118 @@
+    amrex::Real value1, value2, value3, value4, value5, value6, value7, value8;
+    while (file >> value1 >> value2 >> value3 >> value4 >> value5 >> value6 >>
+           value7 >> value8) {
+        Forest f;
+        f.m_type_forest = value1;
+        f.m_x_forest = value2;
+        f.m_y_forest = value3;
+        f.m_height_forest = value4;
+        f.m_diameter_forest = value5;
+        f.m_cd_forest = value6;
+        f.m_lai_forest = value7;
+        f.m_laimax_forest = value8;
+        m_forests.push_back(f);
+    }
+    file.close();
+    m_sim.io_manager().register_output_var("forest_drag");
+    m_sim.io_manager().register_output_var("forest_blank");
+}
+
+void ForestDrag::post_init_actions()
+{
+    BL_PROFILE("amr-wind::" + this->identifier() + "::post_init_actions");
+    const auto& geom_vec = m_sim.repo().mesh().Geom();
+    const int nlevels = m_sim.repo().num_active_levels();
+    for (int level = 0; level < nlevels; ++level) {
+        const auto& geom = geom_vec[level];
+        const auto& dx = geom.CellSizeArray();
+        const auto& prob_lo = geom.ProbLoArray();
+        auto& velocity = m_velocity(level);
+        auto& drag = m_forest_drag(level);
+        auto& blank = m_forest_blank(level);
+        unsigned forestSize = m_forests.size();
+        for (unsigned ii = 0; ii < forestSize; ++ii) {
+            amrex::Real treelaimax = m_forests[ii].calc_lm();
+            const amrex::Real type_forest = m_forests[ii].m_type_forest;
+            const amrex::Real x_forest = m_forests[ii].m_x_forest;
+            const amrex::Real y_forest = m_forests[ii].m_y_forest;
+            const amrex::Real height_forest = m_forests[ii].m_height_forest;
+            const amrex::Real diameter_forest = m_forests[ii].m_diameter_forest;
+            const amrex::Real cd_forest = m_forests[ii].m_cd_forest;
+            const amrex::Real lai_forest = m_forests[ii].m_lai_forest;
+            amrex::Real treeZm = 0.0;
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+            for (amrex::MFIter mfi(velocity); mfi.isValid(); ++mfi) {
+                const auto& vbx = mfi.validbox();
+                auto levelDrag = drag.array(mfi);
+                auto levelBlank = blank.array(mfi);
+                amrex::ParallelFor(
+                    vbx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                        // compute the source term
+                        amrex::Real af = 0;
+                        const amrex::Real x = prob_lo[0] + (i + 0.5) * dx[0];
+                        const amrex::Real y = prob_lo[1] + (j + 0.5) * dx[1];
+                        const amrex::Real z = prob_lo[2] + (k + 0.5) * dx[2];
+                        const amrex::Real radius = std::sqrt(
+                            (x - x_forest) * (x - x_forest) +
+                            (y - y_forest) * (y - y_forest));
+                        if (z <= height_forest &&
```

Comment:
if all this is about computing `af` then you can also make that a member function of `Forest` struct. And then things look really simple: `levelDrag(i, j, k) = cd_forest * f.af(...);`

Reply:
af has a z dependence and we need it inside the lambda.

Reply:
sure, how does it change what I suggested? the af method can take z as an argument.

Reply:
Are we allowed to call a struct from inside the lambda?

Reply:
but you also have the devicevector which contains all the forests on device. So you could avoid a copy to device on every loop by keeping the devicevector. Probably a better way to do it tbh.